### DISCUSSION TITLE
How to handle Collections?

### DISCUSSION THREAD

**Discussion by miwoe** [microstream-one/microstream#451] [Q&A] [answered]
Hi,

I am evaluating Microstream and trying to setup a small app. After some success with some complex types, I am trying now to use Collections to save data sets. First thing, I found out that the standard java implementations use transient fields for the data. Microstreams seems to be unable to save the data in ArrayList, TreeSet etc.

Then I have seen that Microstream has its own Collection package with several implementations, but there seems to be no doc for it. The only api doc class which has a small description is ArrayCollector which is a collection with immutable size or I haven't understood the API, yet.

Can you give me a short kickstart how to use Collections in Microstream? Are there any examples / best practices?

BR,
Michael

**Comment by fh-ms** [comment, marked as answer]
Hi Michael,
MicroStream handles the JDK's collections out of the box. Either with generic handlers or specialized ones.
See the [addendum](https://docs.microstream.one/manual/storage/addendum/specialized-type-handlers.html#_collections) in the documentation.

**Reply by miwoe** [reply]
Hi fh-ms,

thx for the quick answer. I did a retry with JDK's collections and I got it working with these lines:

```
        Collection<Vocab> vocabList = Optional.ofNullable(microStreamConfig.getVocRootStorage().getVocabList()).orElse(new ArrayList<>());
        vocabList.add(newItem);
        microStreamConfig.getVocRootStorage().setVocabList(vocabList);
        microStreamConfig.getStorageManager().store(microStreamConfig.getVocRootStorage().getVocabList());
        microStreamConfig.getStorageManager().store(microStreamConfig.getVocRootStorage());
```
        
So, I ever need to store the collection and then again the root. If I invoke store only for the list, then nothing is stored. If I store only the root, then it seems it stores only the last added item.

I found an example where it seems that only the store-invocation of the list is necessary, actually.

**Reply by fh-ms** [reply]
You have to store the changed objects. In your case, the VocRootStorage is changed by setting the vocabList. Therefore you have to store the root. If you update a previously stored collection, you only have to store the collection of course.

**Reply by miwoe** [reply]
Thx for the detailed explanation!
